src/Http/Connection.php

```diff
@@ -261,4 +265,54 @@ public function setRequestHandler(RequestInterface $requestHandler) : void
     {
         $this->requestHandler = $requestHandler;
     }
+
+    public function getHeaders($requestOptions) : array
+    {
+        if ($this->shouldRefreshToken()) {
+            return $this->tryRefreshHeaderToken();
+        }
+
+         return $requestOptions->getHeaders();
+    }
+
+    public function shouldRefreshToken() : bool
+    {
+        if (isset($this->requestArgs['token']['token_received_on'], $this->requestArgs['token']['expires_in'])) {
+            $token = $this->requestArgs['token'];
+            if ((strtotime('now') - strtotime($token['token_received_on'])) > ($token['expires_in'] - 60)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    public function getNewAccessToken() : string
+    {
+        if (isset($this->requestArgs['token']['clientID'], $this->requestArgs['token']['clientSecret'], $this->requestArgs['token']['outlookProxyUrl'])) {
+            $newToken = $this->getNewToken($this->requestArgs['token']);
+            $this->requestArgs['token']['token_received_on'] = $newToken['token_received_on'];
+            $this->requestArgs['token']['expires_in'] = $newToken['expires_in'];
+            $this->requestArgs['token']['access_token'] = $newToken['access_token'];
+            $this->requestArgs['token']['refresh_token'] = $newToken['refresh_token'];
+
+            return $newToken['access_token'];
+        }
+
+        return '';
+    }
+
+    public function getNewToken(array $token) : array
+    {
+        $tokenObj = new Token($token['clientID'], $token['clientSecret'], ['logger' => $this->logger]);
+        $tokenEntity = $tokenObj->refresh($token['refreshToken'], $token['outlookProxyUrl']);
+        $date = $tokenEntity->tokenReceivedOn();
+
+        return [
```

Comment:
please pass back the tokenEntity itself and use that to hydrate the request args, I see no need to allocate an array and set the request args again

Reply:
Yes, the only reason I did this was to facilitate the testing process with PHPUnit.